extra_data/reader.py

```diff
@@ -623,7 +623,23 @@ def _expand_selection(self, selection):
                     raise SourceNameError(source)
 
                 if keys:
-                    res[source].update(keys)
+                    # keys may be None, which is allowed as an
+                    # undocumented behaviour to support passing a
+                    # DataCollection as the selector.
+
+                    for key in keys:
+                        if not self._has_source_key(source, key):
+                            raise PropertyNameError(key, source)
+
+                    if self.selection[source] is not None:
+                        # If the current DataCollection also selected
+                        # specific keys, make sure the new keys are a
+                        # subset of it.
+                        res[source].update(keys & self.selection[source])
+                    else:
+                        res[source].update(keys)
+                elif self.selection[source] is not None:
+                    res[source] = self.selection[source]
                 else:
                     res[source] = None
```

Comment:
I think this is correct, but it would be easier IMO to follow the code and be sure that all 4 cases (current & new selection limit keys, only current limits, only new limits, neither limit) are handled correctly if it was arranged either as four cases in one if/elif chain, or maybe two nested pairs of branches (which is exactly equivalent to the current code, just not using `elif` as a shortcut).

Or maybe it would be OK with just some clearer comments. I just know that I found it a bit unclear reading the diff as is.

Reply:
Funny, that was my initial implementation, until I realized 2/4 cases are `pass`. I'll gladly add them back to make it explicit.